Checked behavior for NullPointerExceptions.

sjrd opened the pull request:
~~(currently based on #4750 ; only the last commit belongs to this PR)~~

We introduce a new checked behavior `nullPointers`, which covers `NullPointerException`s thrown by IR nodes when one of their arguments is `null`. The relevant nodes are:

* `Apply` and `ApplyStatically` for the receiver,
* `Select` for the qualifier,
* `ArrayLength` and `ArraySelect` for the array,
* `GetClass`, `Clone` and `UnwrapFromException` for their respective only arguments.

In the CoreJSLib, we define a new core function `$n(x)`, which throws when `x === null` and otherwise returns `x`. We give it such a short name because it gets inserted a lot.

We insert a call to `$n` for all the expressions that must not be `null` in the nodes above. We avoid doing so when we can prove that the expression is never `null`. For some simple things, the emitter can locally decide that (`New`, `This`, `LoadModule` and literals are some examples).

The optimizer already had null tracking, notably to elide explicit `null` tests such as those inserted by the compiler for `==`. In order to communicate the enhanced information to the emitter, we use a new transient node `AssumeNotNull(x)`. The emitter can then avoid null checks for those.

We add another transient node `CheckNotNull(x)` for an explicit null check. This is inserted by the optimizer when it expands a node that would other need to perform a null check. The canonical example of that is inlining, which gets rid of the `Apply` node, but must preserve a null check for the receiver. `CheckNotNull` is also used internally by the emitter during its translation.

---

The compiler itself only generated an explicit NPE for the receiver of `synchronized`. We convert that to using a `GetClass(x)` in statement position as a semantics-dependent null check. In theory, this is a breaking change (and that surfaces in changed tests), as the NPE was previously reliable and is not anymore. However, it is very unlikely that anyone would rely on this particular NPE to be thrown.

In general, we use the `x.getClass()`-in-statement-position idiom as a user-space way to perform a semantics-dependent null check. This is semantically accurate, since `getClass()` is pure for all non-null receivers. We make sure that the optimizer can replace it with a `CheckNotNull` or remove it entirely.

---

Benchmarks suggest that the overhead of these new checks in the default fastLink configuration are anywhere from 0% to 17%, with an average of 7%. The code size overhead is of 1.8% in the reversi demo.

It is possible that smarter optimizations could reduce the overhead further. For example, after a successful `x.isInstanceOf[T]`, we could know that `x` is non-null. In fact, we could also know that it is a `T`. Both pieces of knowledge would be relevant to code generated for pattern matches. However, our optimizer is currently not equipped to reason about evaluation-order-dependent knowledge.

---

Overhead of `nullPointers == Fatal` vs `nullPointers == Unchecked`, where the config is otherwise the default `fastLink` config (behaviors are `Fatal` except `moduleInits`, `productionMode == false`, optimizer is enabled, GCC is disabled, ES 2015):

| Benchmark  | Overhead (%) |
|------------|--------------|
| tracer     |          9.9 |
| richards   |         7.20 |
| deltablue  |        16.50 |
| kmeans     |        12.60 |
| sha512     |         6.10 |
| sha512Int  |        14.70 |
| bounce     |         7.00 |
| cd         |        16.70 |
| gcbench    |         3.80 |
| mandelbrot |         1.40 |
| permute    |         2.00 |
| brainfuck  |        10.60 |
| json       |         0.80 |
| list       |         0.00 |
| nbody      |         1.10 |
| queens     |         3.90 |

Average overhead: 7.14%.

sjrd commented:
This is now ready for review. I leave it "draft" because it is based on #4750.

sjrd commented:
I decomposed the giant commit *a little bit*, trying to isolate changes that are not tightly integrated with the NPE checked behavior. Unfortunately, the big commit remains big.

sjrd commented:
I decomposed things a bit further, and even extracted two preparation PRs as #4754 and #4756.

sjrd commented:
Rebased.

@gzm0 Do you want to still have a look at this, or shall I create a Version commit off of `main` now?

gzm0 commented:
> or shall I create a Version commit off of main now?

Let's release. I'll have a look at this now, but it'll take a bit to finish :)

sjrd commented:
Sounds good. There we go: https://github.com/sjrd/scala-js/commit/e1ddd647edb9de2466705b2d8a3b3211132d6ca3

gzm0 commented on partest-suite/src/test/resources/scala/tools/partest/scalajs/2.13.10/run/t8601d.sem:
```diff
@@ -0,0 +1 @@
+nullPointers
```
Did you check that there are no other partests that pass with these semantics? Maybe worth doing in an additional commit.

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/NullPointersTest.scala:
```diff
@@ -0,0 +1,289 @@
+    assertThrows(classOf[NullPointerException], body)
+
+  private def assumeCompliantNPEs(): Unit =
+    assumeTrue("assuming compliant null pointer checks", hasCompliantNullPointers)
```
Annotate this with `@BeforeClass` to make it clear that the entire test assumes NPEs (you can also then remove it from the individual tests).

gzm0 commented on test-suite/shared/src/test/scala/org/scalajs/testsuite/compiler/NullPointersTest.scala:
```diff
@@ -0,0 +1,289 @@
+      assertThrows(classOf[IllegalArgumentException],
+          nullOf[String].substring(throwIllegalArgAsInt()))
+    } else {
+      // charAt is a bit special
```
Maybe clarify that the implementation is special, not the behavior. I tripped over this.

gzm0 reviewed (changes requested):
Mostly nits and minor improvements. One higher level concern about usability of the AssumeNotNull tre.

gzm0 commented:
I have been thinking: Would it be completely outrageous to introduce nullability (and maybe even preciseness) into Scala.js core typesystem?

sjrd commented:
> I have been thinking: Would it be completely outrageous to introduce nullability (and maybe even preciseness) into Scala.js core typesystem?

That would probably be at least somewhat outrageous. :-p

The problem I see with that is that it opens up the IR for all kinds of changes that we would like to do for the sake of the optimizer only. I don't think that's a good idea, as we make the very stable definition dependent on what basically amounts to implementation details.

To be fair, we already crossed that line when we introduced *records* in the IR structure. I'm not sure I can truly call that a *mistake*, as I'm not sure how else we could have reasonably passed them from the optimizer to the emitter otherwise. Nevertheless, it is a bit of a transgression of the principle of separate responsibilities.

gzm0 commented:
I see the argument of not leaking implementation details. But it's not like nullability and/or null-safety is such an arcane concept (exact types feel different) or a concept never provided by compilers (IMO it is fair to open up this to compilers as well).

sjrd commented:
To truly open it up to compilers, we also need to provide `TypeRef`s for non-nullable classes, so that they can be used in `MethodName`s. That's quite some more effort.

If only we were going to use that in Scala 3 with strict nulls, that would be worth it. But the run-time semantics of strict nulls are still going to be erasure-based (like all other semantics in Scala), so realistically we will not be able to compile Scala-level non-nullable types to IR-level non-nullable types without putting portability at risk.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -563,18 +563,22 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
-                constant(arrayTypeRef)
-              case _ =>
-                TailCalls.done(GetClass(finishTransformExpr(texpr)))
+            if (isStat) {
```
Yes, this is necessary to avoid forcing `texpr` (avoid calling `finishTransformExpr(texpr)`) when it is known to be non-null.

sjrd replied:
Hum, actually, with the changes in `keepOnlySideEffects`, this does not seem to have a real impact. I have reverted this particular diff.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -2340,8 +2361,27 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
+         * internal code, we do not do weird things where one of the arguments
+         * is side-effecting as such. And when inlined through the static
+         * method, the arguments have already received Bindings, and are
+         * therefore pure by construction.
```
Indeed, the optimizer can do that for all code outside of the javalib (which must go through the static method). Doing the stupid thing does create a few more intermediate variables for calls inside the javalib, but it is negligible. I think I wrote this code before I came up with `withTempLocalDefs`, and the stupid thing actually required a lot of boilerplate. Now it's easier, so I changed it.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -2462,6 +2504,7 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
             RefinedType(resultType, isExact = isExact, isNullable = false)))
 
       case ArrayBuilderZeroOf =>
+        // This is a private API: `runtimeClass` is known not to be `null`
```
For `GenericArrayBuilderResult`, even with `withNewLocalDefs`, it's very annoying to deal with the null check + the evaluation order + still checking when it's a constant and extracting that. It complicates the code quite a lot, for something that is never going to be necessary. So I would rather not do that.

If we don't do it for `GenericArrayBuilderResult`, it doesn't make sense to do it for `ArrayBuilderZeroOf`.

sjrd commented:
I think I addressed all the comments except checking whether more partest pass with `nullPointers` (I will do that tomorrow)

For the `AssumeNotNull` problem: it is a bit a brittle, yes. But I don't really see what else to do, except duplicating all the NPE-able nodes, as you mentioned, but IMO that would be *more* brittle because of the code duplication and the potential for mistakes. The `AssumeNotNull` should be safe as long as no one intercepts and modifies the IR between the optimizer and the emitter. But that is anyway something that we don't truly support.

sjrd commented on linker/shared/src/test/scala/org/scalajs/linker/OptimizerTest.scala:
```diff
@@ -69,7 +69,7 @@ class OptimizerTest {
 
     def callCloneOn(receiver: Tree): Tree =
-      Apply(EAF, receiver, cloneMethodName, Nil)(AnyType)
+      consoleLog(Apply(EAF, receiver, cloneMethodName, Nil)(AnyType))
```
This is now in the commit that changes `keepOnlySideEffects`.

gzm0 commented:
Last question RE null tracking: Is there anything preventing is from introducing a `TransientType` analogous to the trees? We could then:
- Use this to implement records (IIRC records cannot be transient ATM because they need their own type).
- Use this for null tracking

gzm0 commented:
> We still need to insert AssumeNotNull nodes to actually cast trees into a something with NonNullableType.

Hmmm... If that's the case, then it's indeed pointless. My hope was that the trees we care about here have non-automatic types, so we could just re-type the trees. But indeed, now I realize that we'd need this in the receiver trees. And they are completely arbitrary :-/

gzm0 reviewed (approved):
Unless you object, I propose to go ahead with this and tackle the transient type issue if we see that AssumeNotNull is too brittle.

sjrd commented:
> Unless you object, I propose to go ahead with this and tackle the transient type issue if we see that AssumeNotNull is too brittle.

Sounds good. Let's do that.
